src/com/google/javascript/jscomp/RewriteDynamicImports.java

```diff
@@ -0,0 +1,312 @@
+              UNABLE_TO_COMPUTE_RELATIVE_PATH,
+              importingChunkFilename,
+              targetChunkFilename));
+      return;
+    }
+    final Node originalSpecifierNode = dynamicImport.getFirstChild();
+    Node newSpecifier =
+        astFactory.createString(retargetedSpecifier).srcref(originalSpecifierNode);
+    originalSpecifierNode.replaceWith(newSpecifier);
+    t.reportCodeChange(newSpecifier);
+  }
+
+  /**
+   * Since a dynamic import expression is a promise resolving to the namespace export type, add a
+   * ".then()" call after it and resolve to the rewritten module namespace.
+   *
+   * <p>Before
+   *
+   * <pre>
+   *   import('./foo.js');
+   * </pre>
+   *
+   * After
+   *
+   * <pre>
+   *   import('./foo.js').then(function() { return module$foo; });
+   * </pre>
+   */
+  private void addChainedThen(NodeTraversal t, Node dynamicImport, Var targetModuleNs) {
+    JSTypeRegistry registry = compiler.getTypeRegistry();
+    final Node importParent = dynamicImport.getParent();
+    final Node placeholder = IR.empty();
+    dynamicImport.replaceWith(placeholder);
+    final Node moduleNamespaceNode = createModuleNamespaceNode(targetModuleNs);
+    final Node callbackFn = astFactory.createZeroArgArrowFunctionForExpression(moduleNamespaceNode);
+    final Node importThenCall =
+        astFactory.createCall(astFactory.createGetProp(dynamicImport, "then"), callbackFn);
+    importThenCall
+        .getFirstChild()
+        .setJSType(
+            registry.createFunctionType(
+                registry.createTemplatizedType(
+                    registry.getNativeObjectType(JSTypeNative.PROMISE_TYPE),
+                    registry.getNativeType(JSTypeNative.UNKNOWN_TYPE)),
+                registry.createFunctionType(registry.getNativeType(JSTypeNative.ALL_TYPE))));
+    importThenCall.srcrefTreeIfMissing(dynamicImport);
+    if (dynamicImport.getJSType() != null) {
+      importThenCall.setJSType(dynamicImport.getJSType());
+    }
+    placeholder.replaceWith(importThenCall);
+    compiler.reportChangeToChangeScope(callbackFn);
+    compiler.reportChangeToEnclosingScope(importParent);
+  }
+
+  /**
+   * Replace a dynamic import expression with a function call to the specified alias.
+   *
+   * <p>Before
+   *
+   * <pre>
+   *   import('./foo.js');
+   * </pre>
+   *
+   * After
+   *
+   * <pre>
+   *   aliasedName('./foo.js');
+   * </pre>
+   */
+  private void aliasDynamicImport(NodeTraversal t, Node dynamicImport) {
+    checkNotNull(this.alias);
+    final Node aliasNode = astFactory.createQNameWithUnknownType(this.alias);
+    aliasNode.setOriginalName("import");
+    final Node moduleSpecifier = dynamicImport.getFirstChild().detach();
+    Node importAliasCall =
+        astFactory
+            .createCall(aliasNode, moduleSpecifier)
+            .srcrefTreeIfMissing(dynamicImport);
+    if (dynamicImport.getJSType() != null) {
+      importAliasCall.setJSType(dynamicImport.getJSType());
```

Comment:
@blickly tells me that the preferred way to do this is now `importAliasCall.copyTypeFrom(dynamicImport)`, because this works correctly with the new, simplified "colors" type system we're introducing for use in optimizations.

There are several places where this change needs to be made.

Reply:
My editor is telling me `.copyTypeFrom` is marked as `@deprecated`

Reply:
Changed.

Reply:
Thanks for catching that.  I think it was unintentionally deprecated in https://github.com/google/closure-compiler/tree/f272e4d991f51976f77949d92372178280c11bf4.  I'll send out a change to undeprecate.